packages/core/src/shared/errors.ts

```diff
@@ -886,28 +875,69 @@ function isError(err: Error, id: string, messageIncludes: string = '') {
- * being returned (e.g html). This will cause the client to throw a SyntaxError as a result
+ * AWS SDK clients make requests with the expected result to be JSON data.
+ * But in some cases the request may fail and result in an error HTML page being returned instead
+ * of the JSON. This will cause the client to throw a `SyntaxError` as a result
  * of attempt to deserialize the non-JSON data.
- * While the contents of the response may contain sensitive information, there may be a reason
- * for failure embedded. This function attempts to extract that reason.
  *
- * Example error message before extracting:
- * "Unexpected token '<', "<html><bod"... is not valid JSON Deserialization error: to see the raw response, inspect the hidden field {error}.$response on this object."
+ * But within the `SyntaxError` instance is the real reason for the failure.
+ * This class attempts to extract the underlying issue from the SyntaxError.
  *
- * If the reason cannot be found or the error is not a SyntaxError, return undefined.
+ * Example SyntaxError message before extracting the underlying issue:
+ *  - "Unexpected token '<', "<html><bod"... is not valid JSON Deserialization error: to see the raw response, inspect the hidden field {error}.$response on this object."
+ * Once we extract the real error message from the hidden field `$response.reason` we get messages similar to:
+ *  - "SDK Client unexpected error response: data response code: 403, data reason: Forbidden | Unexpected ..."
  */
-export function getReasonFromSyntaxError(err: Error): string | undefined {
-    if (!(err instanceof SyntaxError)) {
-        return undefined
+export class AwsClientResponseError extends ToolkitError {
+    public static readonly code: string = 'AwsClientResponseError'
+
+    /** Use {@link instanceIf} to create instance. */
+    protected constructor(err: unknown) {
+        const underlyingErrorMsg = AwsClientResponseError.tryExtractReasonFromSyntaxError(err)
+
+        /**
+         * This condition should never be hit since {@link AwsClientResponseError.instanceIf}
+         * is the only way to create an instance of this class, due to the constructor not being public.
+         *
+         * The following only exists to make the types checker happy.
+         */
+        if (!(underlyingErrorMsg && err instanceof Error)) {
+            throw Error(`Cannot create AwsClientResponseError from ${JSON.stringify(err)}}`)
+        }
+
+        super(underlyingErrorMsg, { code: AwsClientResponseError.code, cause: err })
     }
 
-    if (hasKey(err, '$response')) {
-        const response = err['$response']
-        if (response && hasKey(response, 'reason')) {
-            return response['reason'] as string
+    /**
+     * Resolves an instance of {@link AwsClientResponseError} if the given error matches certain criteria.
+     * Otherwise the original error is returned.
+     */
+    static instanceIf<T>(err: T): AwsClientResponseError | T {
```

Comment:
maybe `isFoo` is the usual name for this pattern. example: https://github.com/aws/aws-toolkit-vscode/blob/4c499f5501fff761d45ca17da78084ddafa27233/packages/core/src/shared/errors.ts#L558

that changes the usage pattern but it's not clear why this error needs to have a different usage pattern or why it needs a dedicated class.

```suggestion
    static isAwsClientResponseError<T>(err: T): is AwsClientResponseError | T {
```

Reply:
By not converting the obscure Syntax error to an AwsClientResponseError at the source, we are trusting any downstream code to know to use isFoo to understand the real meaning when they encounter it.

But by creating an AwsClientResponseError right at the source and then throwing that to the downstream, anything that encounters this error will have all the context already in it.

I'd personally rather use an instanceof than a function.

Reply:
Going to assume we do not want `AwsClientResponseError.isClientResponseError()`.
I'll make it `isError()`

Reply:
Also the intent of SDK V3 is to start using subclasses of errors instead so we can use `instanceof`. `isAwsError()` is a byproduct of SDK V2 which we should avoid: https://aws.amazon.com/blogs/developer/service-error-handling-modular-aws-sdk-js/